cpp/modmesh/buffer/SimpleArray.hpp

```diff
@@ -817,6 +819,52 @@ detail::SimpleArrayMixinCalculators<A, T>::median_freq(small_vector<value_type>
     }
 }
 
+/**
+ * Perform matrix multiplication for 2D arrays.
+ * This implementation supports only 2D × 2D matrix multiplication.
+ */
+template <typename A, typename T>
+A SimpleArrayMixinCalculators<A, T>::matmul(A const & other) const
+{
+    auto athis = static_cast<A const *>(this);
+    const size_t this_ndim = athis->ndim();
+    const size_t other_ndim = other.ndim();
+
+    if (this_ndim == 2 && other_ndim == 2)
+    {
+        const size_t m = athis->shape(0);
+        const size_t k = athis->shape(1);
+        const size_t n = other.shape(1);
+
+        if (k != other.shape(0))
+        {
+            throw std::out_of_range(Formatter() << "SimpleArray::matmul(): shape mismatch: this=(" << m << "," << k
+                                                << ") other=(" << other.shape(0) << "," << n << ")");
+        }
+
+        typename detail::SimpleArrayInternalTypes<T>::shape_type result_shape{m, n};
+        A result(result_shape);
+        result.fill(static_cast<value_type>(0));
+
+        for (size_t i = 0; i < m; ++i)
+        {
+            for (size_t j = 0; j < n; ++j)
+            {
+                for (size_t l = 0; l < k; ++l)
+                {
+                    result(i, j) += athis->operator()(i, l) * other(l, j);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    throw std::out_of_range(Formatter() << "SimpleArray::matmul(): unsupported dimensions: this="
```

Comment:
Include the shape of both `this` and `other` matrix, e.g., `(2, 3, 4)` and `(3, 4, 3)`.

Reply:
I think we usually expect to see exception to be thrown in the beginning of the function instead of the end, although either logic is correct.

Please throw/return early if you do not have an argument to do it late.